File: Lexer.py
import Token

class Lexer:
    def __init__(self, input):
        self._input = input
        self._pos = -1
        self._char = ""
        self._peek_char = ""
        self.advance()

    def advance(self):
        self._pos += 1
        self._char = self._input[self._pos] if self._pos < len(self._input) else ""
        self._peek_char = self._input[self._pos + 1] if self._pos + 1 < len(self._input) else ""

    def read_identifier(self):
        result = ""
        while self._char.isalpha() or self._char == "_" :
            result += self._char
            self.advance()
        return result

    def read_number(self):
        result = ""
        while self._char.isdigit():
            result += self._char
            self.advance()
        return result

    def read_token(self):
        while self._char in [" ", "\n", "\r", "\t"]:
            self.advance()
        if not self._char:
            return 0

        # single char tokens
        if self._char in [
            Token.OPEN_PAREN,
            Token.CLOSE_PAREN,
            Token.OPEN_BRACE,
            Token.CLOSE_BRACE,
            Token.FULLSTOP,
            Token.COMMA,
            Token.ASSIGN,
            Token.TO,
            Token.EQUAL_TO,
            Token.THEN,
            Token.WHILE,
            Token.SI,
            Token.ADD,
            Token.MINUS,
            Token.MULTIPLY,
            Token.DIVIDE,
        ]:
            c = self._char
            self.advance()
            return c

        # double char tokens
        elif self._char in [
            Token.INITIALIZE[0],
            Token.LESS_THAN[0],
            Token.GREATER_THAN[0],
            Token.AND[0],
            Token.OR[0],
            Token.NOT[0],
            Token.IF[0],
            Token.GEWA[0],
        ] and self._char + self._peek_char in [
            Token.INITIALIZE,
            Token.LESS_THAN,
            Token.GREATER_THAN,
            Token.AND,
            Token.OR,
            Token.NOT,
            Token.IF,
            Token.GEWA,
        ]:
            tok = self._char + self._peek_char
            self.advance()
            self.advance()
            return tok

        elif self._char.isdigit():
            n = self.read_number()
            return Token.NUMBER

        elif self._char + self._peek_char == "//":
            # commemt, skip until new line
            while not self._char == "\n":
                self.advance()
            return Token.COMMENT

        # else identifier
        i = self.read_identifier()
        if not i:
            self.advance()
            return Token.INVALID
        return Token.IDENTIFIER

File: lexer/lexer.go
package lexer

import (
	token "cantolang/token"
)

type Lexer struct {
	input      []rune
	pos        int
	char       rune
	peekChar   rune
	quotePairs map[rune]rune
}

func New(input string) *Lexer {
	l := &Lexer{
		input: []rune(input),
	}
	l.pos = -1
	l.quotePairs = map[rune]rune{
		'"': '"',
		'“': '”',
		'”': 0,
	}
	l.advance()
	return l
}

func (l *Lexer) advance() {
	l.pos++
	if l.pos < len(l.input) {
		l.char = l.input[l.pos]
	} else {
		l.char = 0
	}
	if l.pos+1 < len(l.input) {
		l.peekChar = l.input[l.pos+1]
	} else {
		l.peekChar = 0
	}
}

func (l *Lexer) readIdentifier() string {
	result := ""
	for isAllowedInIdent(l.char) {
		result += string(l.char)
		l.advance()
	}
	return result
}

func (l *Lexer) readString(endChar rune) string {
	l.advance()
	result := ""
	for l.char != endChar && l.char != 0 {
		result += string(l.char)
		l.advance()
	}
	if l.char == endChar {
		l.advance()
	}
	return result
}

func isAllowedInIdent(char rune) bool {
	if char == 0 {
		return false
	}
	restricted := []rune(" \n1234567890!@#$%^&")
	for _, c := range restricted {
		if c == char {
			return false
		}
	}
	if token.LookUpSymbol(char) == token.TEMP_NOT_SYMBOL {
		return true
	}
	return false
}

func (l *Lexer) readNumber() string {
	result := ""
	for l.char >= '0' && l.char <= '9' {
		result += string(l.char)
		l.advance()
	}
	return result
}

func (l *Lexer) ReadToken() token.Token {
	for l.char == ' ' || l.char == '\n' || l.char == '\r' || l.char == '\t' {
		l.advance()
	}
	if l.char == 0 {
		return token.Token{
			TokenType:    token.EOF,
			TokenLiteral: "",
		}
	}
	var t token.Token

	// check for comment
	if l.char == '/' && l.peekChar == '/' {
		for l.char != '\n' {
			l.advance()
		}
		t.TokenType = token.COMMENT
		return t
	}
	// check for symbol
	symbol := token.LookUpSymbol(l.char)
	if symbol != token.TEMP_NOT_SYMBOL {
		t.TokenType = symbol
		t.TokenLiteral = string(l.char)
		l.advance()
		return t
	}
	// check for number
	if l.char >= '0' && l.char <= '9' {
		t.TokenType = token.NUMBER
		t.TokenLiteral = l.readNumber()
		return t
	}
	// check for string
	matchingQuote, ok := l.quotePairs[l.char]
	if ok {
		if matchingQuote == 0 {
			t.TokenType = token.INVALID
			t.TokenLiteral = string(l.char)
			return t
		}
		t.TokenType = token.STRING
		t.TokenLiteral = l.readString(matchingQuote)
		return t
	}

	// identifier
	i := l.readIdentifier()
	if i == "" {
		t.TokenType = token.INVALID
		t.TokenLiteral = string(l.char)
	} else {
		t.TokenType = token.LookUpIdent(i)
		t.TokenLiteral = i
		return t
	}

	l.advance()
	return t
}
